### PR TITLE
Push orderby and pagination to sql

### PR DESCRIPTION
**Pull request by jcuquemelle**
## What changes are proposed in this pull request?
The ordering and pagination of the requested runs is done in the SQL backend

## How is this patch tested?
Unit Tests
performance test on a dump of a real prod DB: 
- requests.post('....../runs/search', json='{"experiment_ids":["1"]}') 
------ ~14.5 s -> ~5.8s
- requests.post('....../runs/search', json='{"experiment_ids":["1"], "filter":"metrics.triplet_precision >= 0.87"}') ~0.11 s -> ~0.11s
- requests.post('....../runs/search', json='{"experiment_ids":["1"], "max_results": 10, "filter":"attribute.status = \'running\' "}') ~1.32s -> ~0.075s

## Release Notes
performance improvement in the search_runs endpoint

### Is this a user-facing change?
perf only
- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s) does this PR affect?

- [ ] UI
- [ ] CLI
- [ ] API
- [ ] REST-API
- [ ] Examples
- [ ] Docs
- [x] Tracking
- [ ] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 1]
```diff
@@ -781,38 +781,92 @@ def _search(self, experiment_id, filter_string=None,
...
-    def test_order_by_metric(self):
+    def get_ordered_runs(self, order_clauses, experiment_id):
+        return [
+            r.data.tags[mlflow_tags.MLFLOW_RUN_NAME]
+            for r in self.store.search_runs(experiment_ids=[experiment_id],
+                                            filter_string="",
+                                            run_view_type=ViewType.ALL,
+                                            order_by=order_clauses)]
+
+    def test_order_by_metric_tag_param(self):
```
Comment (dbczumar) [inline tests/store/tracking/test_sqlalchemy_store.py]: I tried running this test against a MySQL database, and it emitted the following warning several times:
```
tests/store/tracking/test_sqlalchemy_store.py::TestSqlAlchemyStoreMysqlDb::test_order_by_metric_tag_param
  /Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/sql/base.py:556: SAWarning: Column 'CASE WHEN (anon_1.is_nan IS true) THEN :param_1 WHEN (anon_1.value IS NULL) THEN :param_2 ELSE :param_3 END' on table <sqlalchemy.sql.selectable.Select at 0x1190bebd0; Select object> being replaced by <sqlalchemy.sql.elements.ColumnClause at 0x118e17e10; %(4716925584 anon)s>, which has the same key.  Consider use_labels for select() statements.
    % (key, getattr(existing, "table", None), value)
```

Can you implement the `use_labels` suggestion to remove column name conflicts between tables? This is an argument to `select()` (https://docs.sqlalchemy.org/en/13/core/selectable.html#sqlalchemy.sql.expression.select).

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 2]
```diff
@@ -687,3 +699,60 @@ def _get_sqlalchemy_filter_clauses(parsed, session):
...
+                order_value = subquery.c.value
+
+            # sqlite does not support NULLS LAST expression, so we sort first by
+            # presence of the field (and is_nan for metrics), then by actual value
+            if SearchUtils.is_metric(key_type, '='):
+                clauses.append(sql.case([
+                    (subquery.c.is_nan.is_(True), 1),
+                    (order_value.is_(None), 1)
+                ],
+                    else_=0))
```
Comment (dbczumar) [inline mlflow/store/tracking/sqlalchemy_store.py]: Nit: The formatting here looks a bit odd. Can we move the `else` kwarg up to the preceding line to read `], else=0))` ?

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 3]
```diff
@@ -781,38 +781,92 @@ def _search(self, experiment_id, filter_string=None,
...
+        # desc / desc
+        self.assertListEqual(["inf/2", "1000/7", "0/6", "0/5", "-1000/4", "-inf/3", "nan/1"],
+                             self.get_ordered_runs(["metrics.x desc", "param.metric desc"],
+                                                   experiment_id))
+
+    def test_order_by_attributes(self):
+        experiment_id = self.store.create_experiment('order_by_attributes')
+
+        def create_run(start_time, end):
+
```
Comment (dbczumar) [inline tests/store/tracking/test_sqlalchemy_store.py]: Nit: Remove blank line

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 4]
```diff
@@ -1408,29 +1462,42 @@ def test_search_runs_returns_expected_results_with_large_experiment(self):
...
 
         run_results = self.store.search_runs([experiment_id],
                                              "metrics.mkey_0 < 26 and metrics.mkey_0 > 5 "
                                              "and tags.tkey_0 = 'tval_0' "
                                              "and params.pkey_0 = 'pval_0'",
-                                             ViewType.ALL, max_results=1000)
+                                             ViewType.ALL, max_results=5)
         assert len(run_results) == 1  # 2 runs on previous request, 1 of which has a 0 pkey_0 value
 
+    def test_get_attribute_name(self):
```
Comment (dbczumar) [inline tests/store/tracking/test_sqlalchemy_store.py]: Can you merge this test with the pre-existing `test_get_attribute_name()` test that you introduced in the filter pushdown PR? This test is not specific to sqlite backends.

**Review by dbczumar — CHANGES_REQUESTED** [review]
@jcuquemelle This is a fantastic improvement to MLflow SQLAlchemy / UI performance. I left a few small comments, nothing major. It's almost ready!

**Review by dbczumar — CHANGES_REQUESTED** [review]
@jcuquemelle Thanks for addressing the last round of comments.

When running tests against MySQL, I still see the following output for certain search queries:
```
/Users/czumar/anaconda2/lib/python2.7/site-packages/sqlalchemy/sql/base.py:556: SAWarning: Column 'CASE WHEN (anon_1.latest_metrics_is_nan IS true) THEN :param_1 WHEN (anon_1.latest_metrics_value IS NULL) THEN :param_2 ELSE :param_3 END' on table <sqlalchemy.sql.selectable.Select at 0x1102e4e90; Select object> being replaced by <sqlalchemy.sql.elements.ColumnClause at 0x1102e4690; %(4566185680 anon)s>, which has the same key.  Consider use_labels for select() statements.
  % (key, getattr(existing, "table", None), value)
```

This appears to occur when executing a search query with *two or more* orderby clauses referencing the *same* table (`metrics`, `tags`, `params`, etc). Here's some example code that I used to reproduce the issue:

```
import mlflow
mlflow.set_tracking_uri("<MYSQL_DB_TRACKING_URI>")

for i in range(10):
    with mlflow.start_run():
        mlflow.log_param("paramA", i)
        mlflow.set_tag("tagA", i)
        mlflow.log_metric("metricA", i)
        mlflow.log_metric("metricB", 10 - i)

runs = mlflow.search_runs("0", order_by=["metrics.metricA asc", "metrics.metricB asc"]) # Raises warnings
```

This error doesn't seem to occur when the orderby clauses reference *distinct* tables. Can you take a look and see what might be going on?

**Comment by jcuquemelle** [comment]
> This error doesn't seem to occur when the orderby clauses reference _distinct_ tables. Can you take a look and see what might be going on?

The issue was that the name conflict was in fact occuring between all CASE WHEN columns, using labels on each subquery would solve between-tables conflicts only. The new solution reverts labelling columns in the subquery but assigns a unique label for each CASE WHEN clause

**Comment by jcuquemelle** [comment]
close and reopen to relaunch Travis (got a iris_training.csv: 503 -- Service Unavailable error in py 2.7 tests)

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 5]
```diff
@@ -1305,18 +1358,18 @@ def test_metrics_materialization_upgrade_succeeds_and_produces_expected_latest_m
...
-        >>> mlflow.set_tracking_uri(
-        ...     "sqlite:///../../resources/db/db_version_7ac759974ad8_with_metrics.sql")
-        >>> client = MlflowClient()
-        >>> summary_metrics = {
-        ...     run.info.run_id: run.data.metrics for run
-        ...     in client.search_runs(experiment_ids="0")
-        ... }
-        >>> with open("dump.json", "w") as dump_file:
-        >>>     json.dump(summary_metrics, dump_file, indent=4)
+        # >>> import json
```
Comment (dbczumar) [inline tests/store/tracking/test_sqlalchemy_store.py]: Nit: @jcuquemelle Can we uncomment this code?

**Review by dbczumar — APPROVED** [review]
@jcuquemelle LGTM! This is awesome! I left one small comment about uncommenting test docs; please let me know if this can be addressed

File: tests/store/tracking/test_sqlalchemy_store.py (hunk 5 above)
Comment (jcuquemelle) [reply tests/store/tracking/test_sqlalchemy_store.py]: of course ... forgot this shouldn't have been commited !

**Pull request merged**